### PR TITLE
Smaller (in-memory) and lazy edge lists

### PR DESCRIPTION
**Pull request by johandahlberg**
## Description

First of all, sorry that this is such a huge PR, but this touches on a lot different things.

This introduces:
 - smaller in memory edge lists (~60% of the original dataframe size in my benchmarks)
 - less memory intensive concatenation (~23% of the original memory usage)
 - lazy data frame representation of the edge list using polars `LazyFrame`. This allows us to carry out many operations on the edge list without loading all of it into memory, which should improve scalability in many cases.
   - the lazy data frames are now used, for checking the edge list size, filtering `PixelDataset` instances, concatenation, and finding individual components in the edgelist

These changes are backward compatible, but there will be a performance hit (higher memory usage, and longer runtimes) when working with .pxl-files generated with older pixelator versions.

Fixes: https://linear.app/pixelgen-technologies/issue/EXE-1062/reduce-edgelist-memory-usage-in-concatenation

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

This has been tested by manually running concatenation, and manually trying out some of these operations in a quarto notebook.

## PR checklist:

- [ ] This comment contains a description of changes (with reason).
- [ ] My code follows the style guidelines of this project
- [ ] Make sure your code lints, is well-formatted and type checks
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If you've added a new stage - have you followed the pipeline CLI conventions in the [USAGE guide](../USAGE.md)
- [ ] [README.md](./README.md) is updated
- [ ] If a new tool or package is included, update poetry.lock, [the conda recipe](../conda-recipe/pixelator/meta.yaml) and [cited it properly](../CITATIONS.md)
- [ ] Include any new [authors/contributors](../AUTHORS.md)
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and documentation and corrected any misspellings
- [ ] Usage Documentation is updated
- [ ] If you are doing a [release](../RELEASING.md#Releasing), or a significant change to the code, update [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: src/pixelator/pixeldataset.py
Comment (ambarrio) [inline src/pixelator/pixeldataset.py:68]: I think the reuse of `component` here is a bit confusing between the keyword component we use for not-yet called cells and for the components of the pxl structure as I guess this refers to. 

Can we change the method? Can we standarize how we call things? In the technical documentation pertaining the PXL file, we call them elements.

File: src/pixelator/pixeldataset.py
Comment (ambarrio) [inline src/pixelator/pixeldataset.py:583]: It doesn't default to `None`. It is just optional

Comment (ambarrio) [reply src/pixelator/pixeldataset.py:583]: Should we just delete the defaults to None if we need it not to?

**Review by ambarrio — COMMENTED** [review]
I sent you my comments, we already talked a bit over Slack. Let's resolve them and merge on Monday perhaps. So we can also try the 40M limit as well.

File: src/pixelator/analysis/colocalization/prepare.py
Comment (johandahlberg) [reply src/pixelator/analysis/colocalization/prepare.py:81]: Yeah, it means `marker > min_marker_counts`. I can change it.

File: src/pixelator/graph/utils.py
Comment (johandahlberg) [reply src/pixelator/graph/utils.py:59]: If we set `observed=False` all factors will be used, i.e. even those that are not present the current grouping. To preserve the behavior of this function, we therefore need to set `observed=True`.

File: src/pixelator/graph/utils.py
Comment (johandahlberg) [reply src/pixelator/graph/utils.py:137]: Good catch!

File: tests/analysis/colocalization/test_prepare.py
Comment (johandahlberg) [reply tests/analysis/colocalization/test_prepare.py:65]: I think I did this before we had merged the config changes that remove the need to add docstring in the tests. I will remove it now.

File: tests/test_pixeldataset.py
Comment (johandahlberg) [reply tests/test_pixeldataset.py]: So, the reason for this is that we were using the `pyarrow` extension when reading these data frames (and then not converting them). I dropped this now, since especially for these smaller data frames it doesn't make that much of a difference in read performance/memory usage. We might go back to using `pyarrow` more extensively in the future, but I think we should wait for the rest of the ecosystem to catch up first.

File: tests/test_pixeldataset.py
Comment (johandahlberg) [reply tests/test_pixeldataset.py:867]: I was debugging a thing here and forgot to remove it. I'll clean it up.

File: src/pixelator/pixeldataset.py
Comment (johandahlberg) [reply src/pixelator/pixeldataset.py:68]: No, this refers to this part:
```
            attr["component"] = pd.Categorical(
                attr["component"].astype(str) + "_" + attr["sample"].astype(str)
            )
            attr.set_index(["component"])
```

That incorporates the sample into the component column (to make it unique), and set's it as the index of the data frame. So I actually think `component` in the name of this function makes sense.

File: src/pixelator/pixeldataset.py
Comment (johandahlberg) [reply src/pixelator/pixeldataset.py:581]: Yes. Good catch. I just missed setting this back since I went a bit back and forth in the implementation of this.

**Review by ambarrio — APPROVED** [review]
I leave my comments but I think this is ready to go actually, so you have my approval.

The only comment to mark is the sorting when comparing frames

File: tests/test_pixeldataset.py
Comment (johandahlberg) [reply tests/test_pixeldataset.py:28]: This was a pragmatic decision on my part here. In general we aren't testing private methods. But in this case I thought it would useful since it's a rather crucial part of data transformation. I'll add a comment about this in the test to explain my reasoning.

File: tests/graph/test_graph_utils.py
Comment (johandahlberg) [reply tests/graph/test_graph_utils.py:166]: Some of the methods do not guarantee that the index will be returned sorted, so for tests to work I sort them. I'll make a comment about in the code.

**Pull request merged**
